verilog/tools/kythe/kythe_facts_extractor.h

```diff
@@ -0,0 +1,109 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VERIBLE_VERILOG_TOOLS_KYTHE_KYTHE_FACTS_EXTRACTOR_H_
+#define VERIBLE_VERILOG_TOOLS_KYTHE_KYTHE_FACTS_EXTRACTOR_H_
+
+#include <string>
+#include <utility>
+
+#include "verilog/tools/kythe/indexing_facts_tree.h"
+#include "verilog/tools/kythe/indexing_facts_tree_context.h"
+#include "verilog/tools/kythe/indexing_facts_tree_extractor.h"
+#include "verilog/tools/kythe/kythe_facts.h"
+
+namespace verilog {
+namespace kythe {
+
+// Streamable printing class for kythe facts.
+// Usage: stream << KytheFactsPrinter(*tree_root);
+class KytheFactsPrinter {
+ public:
+  explicit KytheFactsPrinter(const IndexingFactNode& root) : root_(root) {}
+
+  std::ostream& Print(std::ostream&) const;
+
+ private:
+  // The root of the indexing facts tree to extract kythe facts from.
+  const IndexingFactNode& root_;
+};
+
+// Responsible for traversing IndexingFactsTree and processing its different
+// nodes to produce kythe indexing facts.
+class KytheFactsExtractor {
+ public:
+  explicit KytheFactsExtractor(absl::string_view file_path,
+                               std::ostream* stream)
+      : file_path_(file_path), stream_(stream) {}
+
+  void Visit(const IndexingFactNode&);
+
+ private:
+  // Container with a stack of VNames to hold context of VNames during traversal
+  // of an IndexingFactsTree.
+  class VNameContext : public verible::AutoPopStack<const VName*> {
+   public:
+    typedef verible::AutoPopStack<const VName*> base_type;
+
+    // member class to handle push and pop of stack safely
+    using AutoPop = base_type::AutoPop;
+
+   public:
+    // returns the top VName of the stack
+    const VName& top() const { return *ABSL_DIE_IF_NULL(base_type::top()); }
+  };
+
+  // Extracts kythe facts from file node and returns it VName.
+  VName ExtractFileFact(const IndexingFactNode&);
+
+  // Extracts kythe facts from module instance node and returns it VName.
+  VName ExtractModuleInstanceFact(const IndexingFactNode&);
+
+  // Extracts kythe facts from module node and returns it VName.
+  VName ExtractModuleFact(const IndexingFactNode&);
+
+  // Extracts kythe facts from module port node and returns its VName.
+  VName ExtractVariableDefinition(const IndexingFactNode& node);
+
+  // Extracts kythe facts from a module port reference node and returns its
+  // VName.
+  VName ExtractVariableReference(const IndexingFactNode& node);
+
+  // Generates an anchor VName for kythe.
+  VName PrintAnchorVName(const Anchor&, absl::string_view);
+
+  // The verilog file name which the facts are extracted from.
+  std::string file_path_;
```

Comment:
if this never changes and is always initialized during construction, it can be a `const` member.